src/common/StackStringStream.h

```diff
@@ -30,43 +30,56 @@ template<std::size_t SIZE>
 class StackStringBuf : public std::basic_streambuf<char>
 {
 public:
-  StackStringBuf() = default;
+  StackStringBuf()
+    : vec{SIZE, boost::container::default_init_t{}}
+  {
+    setp(vec.data(), vec.data() + vec.size());
+  }
   StackStringBuf(const StackStringBuf&) = delete;
   StackStringBuf& operator=(const StackStringBuf&) = delete;
   StackStringBuf(StackStringBuf&& o) = delete;
   StackStringBuf& operator=(StackStringBuf&& o) = delete;
   ~StackStringBuf() override = default;
 
-  void push(std::string_view sv)
-  {
-    vec.insert(vec.end(), sv.begin(), sv.end());
-  }
-
   void clear()
   {
-    vec.clear();
+    vec.resize(SIZE);
+    setp(vec.data(), vec.data() + SIZE);
   }
 
   std::string_view strv() const
   {
-    return std::string_view(vec.data(), vec.size());
+    return std::string_view(pbase(), pptr() - pbase());
   }
 
 protected:
   std::streamsize xsputn(const char *s, std::streamsize n)
   {
-    push(std::string_view(s, n));
+    std::streamsize capacity = epptr() - pptr();
+    std::streamsize left = n;
+    if (capacity >= left) {
+      memcpy(pptr(), s, left);
```

Comment:
@branch-predictor could you enlighten me why `maybe_inline_memcpy()` might outperform `memcpy()` in this case? as `capacity` is not a constant here anymore after the buffer is overflowed. and in that case, the implementation[0] of `memcpy()` is quite the same as what we have.

---
[0] https://github.com/lattera/glibc/blob/master/string/memcpy.c#L27

Reply:
@tchaikov Sure. ```memcpy``` / ```memmove``` will be inlined (no function call generated) if number of bytes to copy is constant (known upfront), but here it's not because that's a variable. ```maybe_inline_memcpy``` will be always inlined. Plus, if amount of data to copy is 1,2 3, 4 or 8 bytes, it will be done in single step without any loops.

Reply:
@branch-predictor sorry, i was wrong. the implementation of `memcpy()` is https://github.com/lattera/glibc/blob/a2f34833b1042d5d8eeb263b4cf4caaea138c4ad/sysdeps/x86_64/memcpy.S#L56  (on amd64). i think it also copies without unnecessary loops.

so i think the only advantage of `maybe_inline_memcpy()` is that it is inlined.

Reply:
@tchaikov The function you linked is first copying first 1, 2, 4 and 8 bytes (if applicable) and then proceeds with loop to copy remaining bytes (if the block is <= 31 bytes large), if the block size is equal to 1, 2, 4 or 8 bytes, it still checks for other sizes (contrary to ```maybe_inline_memcpy```), so the advantage is not just it being inlined, but also less checks (though the difference is not going to be shocking in any way). The main advantage remains it being inlined, yes.

Reply:
> if the block size is equal to 1, 2, 4 or 8 bytes, it still checks for other sizes

yes, i think the reason is that it tries to use movq in larger batch. while `maybe_inline_memcpy()` simply calls the inlined version of `__builtin_memcpy()` which probably does not try that hard. i failed to find/understand how the block size is calculated in GCC. probably it's https://github.com/gcc-mirror/gcc/blob/master/gcc/builtins.c#L3223 . i don't have enough bandwidth reading/understanding it. guess GCC will do it right.

i also tried to understand the distribution of line width of logging in OSD.

```
 (108, 28),
 (106, 30),
 (113, 30),
 (104, 31),
 (105, 32),
 (88, 36),
 (89, 36),
 (102, 38),
 (101, 60),
 (100, 312)
```
above is histogram of the osd log from a vstart cluster. where there is the 100-chars log lines were printed 312 times. so i guess we need to optimize for the case where the buffer size is 100. and i *guess*, each call of `xsputn()` should put `3` ~ `30` characters in average. so presumably, the `switch` clause can help us to take care of quite a few `xsputn()` calls with small `n`.

@branch-predictor sorry for making a fuss on this, as i try to trust the glibc and to avoid use a function without understanding its impact. as IMHO, `memcpy()` is a relatively safe bet. IIRC, `std::memmove()` is  used to implement the `boost::vector::insert()`. so existing change should be a safe move. but with your help, i think `maybe_inlined_memcpy()` is better. thank you.